pwiz_tools/Skyline/FileUI/PeptideSearch/EncyclopeDiaSearchDlg.cs

```diff
@@ -599,27 +603,80 @@ public IFormView ShowingFormView
 
     public class EncyclopeDiaSearchControl : SearchControl
     {
-        public EncyclopeDiaSearchControl(Control hostControl)
+        public EncyclopeDiaSearchDlg HostDialog { get; }
+
+        public EncyclopeDiaSearchControl(EncyclopeDiaSearchDlg hostControl)
         {
-            Parent = hostControl;
+            Parent = HostDialog = hostControl;
         }
 
         public EncyclopeDiaSearchDlg.EncyclopeDiaSettings Settings { get; set; }
         public string EncyclopeDiaChromLibraryPath { get; private set; }
         public string EncyclopeDiaQuantLibraryPath { get; private set; }
 
+        public class ParallelRunnerProgressControl : MultiProgressControl, IProgressMonitor
+        {
+            private readonly EncyclopeDiaSearchControl _hostControl;
+
+            public ParallelRunnerProgressControl(EncyclopeDiaSearchControl hostControl)
+            {
+                _hostControl = hostControl;
+                ProgressSplit.Panel2Collapsed = true;
+            }
+
+            // ReSharper disable once InconsistentlySynchronizedField
+            public bool IsCanceled => _hostControl.IsCanceled;
+
+            public UpdateProgressResponse UpdateProgress(IProgressStatus status)
+            {
+                if (IsCanceled || status.IsCanceled)
+                    return UpdateProgressResponse.cancel;
+
+                var match = Regex.Match(status.Message, @"(.*)\:\:(.*)");
+                Assume.IsTrue(match.Success && match.Groups.Count == 3,
+                    @"ParallelRunnerProgressDlg requires a message like file::message to indicate which file's progress is being updated");
+
+                lock(this)
+                {
+                    // only make the MultiProgressControl visible if it's actually used
+                    if (RowCount == 0)
+                    {
+                        var hostDialog = _hostControl.HostDialog;
+                        if (hostDialog.InvokeRequired)
```

Comment:
I think you're right that it should just always Invoke/BeginInvoke here because this UpdateProgress is basically always going to be called from a worker thread.

I don't see how it could deadlock. The point of the lock is to avoid the dialog widening code running more than once. I could switch to BeginInvoke just in case?